File: routes/payments.js
const express = require('express');
const Razorpay = require('razorpay');
const crypto = require('crypto');
const fs = require('fs').promises;
const router = express.Router();
const {
  cleanupFile,
  createUploadMiddleware,
  dispatchPrintJob,
  maxUploadSize
} = require('./print-helper');

const upload = createUploadMiddleware();

if (!process.env.RAZORPAY_KEY_ID || !process.env.RAZORPAY_KEY_SECRET) {
  console.warn("Razorpay API keys are not configured. Payment will not work.");
}

let razorpay = null;
if (process.env.RAZORPAY_KEY_ID && process.env.RAZORPAY_KEY_SECRET) {
  razorpay = new Razorpay({
    key_id: process.env.RAZORPAY_KEY_ID,
    key_secret: process.env.RAZORPAY_KEY_SECRET,
  });
}

// Create Razorpay order
router.post('/create-razorpay-order', async (req, res) => {
  try {
    const { amount, piId, printerId, pages, userId } = req.body;

    if (!amount || !piId || !printerId || !pages || !userId) {
      return res.status(400).json({ 
        error: 'Missing or invalid required parameters.',
        details: 'amount, piId, printerId, pages, and userId are required'
      });
    }

    if (!razorpay) {
      return res.status(500).json({ error: 'Payment gateway is not configured on the server.' });
    }

    const amountInPaise = Math.round(amount * 100);

    const options = {
      amount: amountInPaise,
      currency: 'INR',
      receipt: `receipt_print_${new Date().getTime()}`,
      notes: {
        userId,
        piId,
        printerId,
        pages,
      }
    };

    const order = await razorpay.orders.create(options);
    res.json({ order_id: order.id, finalAmount: amount });
  } catch (error) {
    res.status(500).json({ error: 'Failed to create payment order.' });
  }
});

function verifyRazorpaySignature(orderId, paymentId, signature) {
  if (!process.env.RAZORPAY_KEY_SECRET) {
    throw new Error('Payment gateway is not configured on the server.');
  }

  const expectedSignature = crypto
    .createHmac('sha256', process.env.RAZORPAY_KEY_SECRET)
    .update(`${orderId}|${paymentId}`)
    .digest('hex');

  return expectedSignature === signature;
}

function hasPrintPayload(req) {
  const { piId, printerId, userId } = req.body || {};
  return Boolean(req.file && piId && printerId && userId);
}

// Verify payment and optionally dispatch the print immediately.
router.post('/verify', upload.single('file'), async (req, res) => {
  try {
    const { paymentId, orderId, signature } = req.body;

    if (!paymentId || !orderId || !signature) {
      if (req.file) {
        cleanupFile(req.file.path);
      }
      return res.status(400).json({ error: 'paymentId, orderId, and signature are required.' });
    }

    const signatureValid = verifyRazorpaySignature(orderId, paymentId, signature);
    if (!signatureValid) {
      if (req.file) {
        cleanupFile(req.file.path);
      }
      return res.status(400).json({ error: 'Invalid payment signature.' });
    }

    const payment = {
      paymentId,
      orderId,
      signature,
      signatureVerified: true,
      status: 'completed',
      createdAt: new Date()
    };

    let printResult = null;
    if (hasPrintPayload(req)) {
      printResult = await dispatchPrintJob(req);
      payment.printTriggered = true;
      payment.printTrackingId = printResult.trackingId;
      payment.printJobId = printResult.jobId;
    } else {
      payment.printTriggered = false;
    }

    await req.db.collection('payments').insertOne(payment);

    if (req.file) {
      try {
        await fs.unlink(req.file.path);
      } catch (cleanupError) {
        console.error('[Payment Verify] Failed to cleanup local file:', cleanupError);
      }
    }

    res.json({
      success: true,
      verified: true,
      printTriggered: Boolean(printResult),
      ...(printResult ? { print: printResult } : {})
    });
  } catch (error) {
    if (req.file) {
      cleanupFile(req.file.path);
    }

    let errorMessage = error.message;
    if (error.code === 'LIMIT_FILE_SIZE') {
      errorMessage = `File too large. Maximum allowed size is ${Math.floor(maxUploadSize / (1024 * 1024))}MB.`;
    }

    res.status(error.statusCode || 500).json({ error: errorMessage });
  }
});

module.exports = router;

File: routers/payments.py
from fastapi import APIRouter, HTTPException, Depends
from pydantic import BaseModel
import razorpay
import os
from datetime import datetime
from pymongo.database import Database

router = APIRouter()

# Initialize Razorpay client
razorpay_client = None
razorpay_key_id = os.getenv('RAZORPAY_KEY_ID')
razorpay_key_secret = os.getenv('RAZORPAY_KEY_SECRET')

if razorpay_key_id and razorpay_key_secret:
    razorpay_client = razorpay.Client(auth=(razorpay_key_id, razorpay_key_secret))
    print("Razorpay configured successfully")
else:
    print("Razorpay API keys are not configured. Payment will not work.")

# Pydantic models
class CreateOrder(BaseModel):
    amount: float
    piId: str
    printerId: str
    pages: int
    userId: str

class VerifyPayment(BaseModel):
    paymentId: str
    orderId: str
    signature: str

# Dependency to get database
def get_database():
    from main import db
    if db is None:
        raise HTTPException(status_code=500, detail="Database not connected")
    return db

@router.post("/create-razorpay-order")
async def create_razorpay_order(order: CreateOrder, db: Database = Depends(get_database)):
    try:
        if not razorpay_client:
            raise HTTPException(status_code=500, detail="Payment gateway is not configured on the server.")

        amount_in_paise = round(order.amount * 100)

        options = {
            "amount": amount_in_paise,
            "currency": "INR",
            "receipt": f"receipt_print_{int(datetime.now().timestamp())}",
            "notes": {
                "userId": order.userId,
                "piId": order.piId,
                "printerId": order.printerId,
                "pages": str(order.pages),
            }
        }

        razorpay_order = razorpay_client.order.create(data=options)
        return {"order_id": razorpay_order["id"], "finalAmount": order.amount}
    except HTTPException:
        raise
    except Exception as e:
        print(f"Payment order creation error: {str(e)}")
        raise HTTPException(status_code=500, detail=f"Failed to create payment order: {str(e)}")

@router.post("/verify")
async def verify_payment(payment: VerifyPayment, db: Database = Depends(get_database)):
    try:
        payment_doc = {
            "paymentId": payment.paymentId,
            "orderId": payment.orderId,
            "signature": payment.signature,
            "status": "completed",
            "createdAt": datetime.now()
        }
        
        db.payments.insert_one(payment_doc)
        return {"success": True}
    except Exception as e:
        print(f"Payment verification error: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))